### ISSUE TITLE
[BUG] - A small subset of nodes on alonzo-purple do not respond to new connections

### ISSUE THREAD

**Issue by rdlrt** [input-output-hk/cardano-node#3080]
**External**

**Other**

**Summary**

At some point (different times individually) post node startup , multiple folks from alonzo-purple network have reported an issue where their nodes are not responding to a new connection, be it via cardano-ping or a simple netcat, the connections end up with a timeout. Node in question for this instance is "178.238.224.187:6003".

- The issue is not related to build steps (in this case, I used hydra link to download binaries - but others have experienced it using `cabal build` as well).
- The topology.json, config.json  used were exactly as per defaults from [cardano-deployment job](https://hydra.iohk.io/build/7189190/download/1/index.html), only deviation from default for me (not for others) was changing output logging to file and turning TraceMempool to False.

**Steps to reproduce**
1. Start node
2. After a while, try netcat/cardano-ping against node.

**Expected behavior**
The node should respond to the connection

**System info (please complete the following information):**
- OS Name: CentOS
- OS Version: 8
- Node version: 1.28.0 - linux-x86_64 - ghc-8.10, 708de685d49ec6af4b2d8b3cbfa0eca0e9e43edf
- CLI version : 1.28.0 - linux-x86_64 - ghc-8.10, 708de685d49ec6af4b2d8b3cbfa0eca0e9e43edf

**Screenshots and attachments**
- netstat against node port (as asked by Sam)
![image](https://user-images.githubusercontent.com/3169068/129467322-8022f782-0c04-4ec4-ab0c-a3a1ada14041.png)
- (Non-Info level) log extract [here](https://gist.githubusercontent.com/rdlrt/d90345b84bb855f6bd4ccb4173c7173f/raw/618408b080e47187baf942c32d850a62af43c6b9/cnode_info.log), with `AddedTo` messages stripped

**Additional context**
- No relevant error in default logging from node
- No killed processes/dmesg errors since node uptime
- No thread in T or Z state as per OS
- Restarting node will likely resolve the issue, but have kept node on to be able to test/provide any information that could be useful.

**Comment by coot** [comment]
The `NetworkMagic` does not match:
```
ErrorPolicyUnhandledApplicationException
  (Refused NodeToNodeV_7
    \"version data mismatch: NodeToNodeVersionData {networkMagic = NetworkMagic {unNetworkMagic = 8}, diffusionMode = InitiatorAndResponderDiffusionMode}
                          /= NodeToNodeVersionData {networkMagic = NetworkMagic {unNetworkMagic = 7}, diffusionMode = InitiatorAndResponderDiffusionMode}\")
```
it might just be a misconfiguration.

**Comment by rdlrt** [comment]
Those are incoming connections @coot , not a misconfiguration - but old nodes from alonzo white who did not update

**Comment by gitmachtl** [comment]
i ran a test: when a node responds to a ping, i directly reping them 10 times in a row. as you can see, most of the nodes are responding with around the same RTT as with the initial test ping, like the iog relays. but than we have nodes like HOMR1 with a fast initial response, but a pretty lame respond after that, also with increasing numbers. same as with YOADA, starting fast, after than jumping directly into the 4second RTT range. the node TUNDR has a rate-limit set to not allow more than 5 incoming connections from the same host, that is working like it should.

maybe this can be a hint where to start searching?

![image](https://user-images.githubusercontent.com/47434720/129473496-7fd275f2-8b43-4727-925d-b68af8f4d323.png)

**Comment by coot** [comment]
It seems fine now for `HOMR1`, `YADA` and `TUNDR`:

```
cardano-ping -c10 -m 8 -h 95.217.220.132 -p 6000   
95.217.220.132:6000 network rtt: 0.075
95.217.220.132:6000 handshake rtt: 0.072039675s
95.217.220.132:6000 Negotiated version NodeToNodeVersionV7 8 False
                           timestamp,                 host,  cookie,  sample,  median,     p90,    mean,     min,     max,     std
    2021-08-15 10:07:31.12251605 UTC,  95.217.220.132:6000,       0,   0.082,   0.082,   0.082,   0.082,   0.082,   0.082,     NaN
   2021-08-15 10:07:32.187829081 UTC,  95.217.220.132:6000,       1,   0.063,   0.073,   0.080,   0.073,   0.063,   0.082,   0.013
   2021-08-15 10:07:33.281678624 UTC,  95.217.220.132:6000,       2,   0.092,   0.080,   0.091,   0.079,   0.063,   0.092,   0.015
    2021-08-15 10:07:34.35098461 UTC,  95.217.220.132:6000,       3,   0.068,   0.075,   0.090,   0.076,   0.063,   0.092,   0.013
   2021-08-15 10:07:35.422744578 UTC,  95.217.220.132:6000,       4,   0.070,   0.073,   0.090,   0.075,   0.063,   0.092,   0.012
   2021-08-15 10:07:36.491237571 UTC,  95.217.220.132:6000,       5,   0.067,   0.069,   0.089,   0.074,   0.063,   0.092,   0.011
   2021-08-15 10:07:37.559689391 UTC,  95.217.220.132:6000,       6,   0.068,   0.068,   0.089,   0.073,   0.063,   0.092,   0.010
    2021-08-15 10:07:38.63706605 UTC,  95.217.220.132:6000,       7,   0.076,   0.069,   0.088,   0.073,   0.063,   0.092,   0.010
   2021-08-15 10:07:39.717343161 UTC,  95.217.220.132:6000,       8,   0.079,   0.070,   0.088,   0.073,   0.063,   0.092,   0.009
   2021-08-15 10:07:40.797313284 UTC,  95.217.220.132:6000,       9,   0.079,   0.072,   0.087,   0.074,   0.063,   0.092,   0.009


cardano-ping -c 10  -m 8 -h 217.182.129.124 -p 6000
217.182.129.124:6000 network rtt: 0.045
217.182.129.124:6000 handshake rtt: 0.052140451s
217.182.129.124:6000 Negotiated version NodeToNodeVersionV7 8 False
                           timestamp,                 host,  cookie,  sample,  median,     p90,    mean,     min,     max,     std
   2021-08-15 10:07:04.382866093 UTC, 217.182.129.124:6000,       0,   0.056,   0.056,   0.056,   0.056,   0.056,   0.056,     NaN
   2021-08-15 10:07:05.429563616 UTC, 217.182.129.124:6000,       1,   0.045,   0.051,   0.055,   0.051,   0.045,   0.056,   0.008
   2021-08-15 10:07:06.493191718 UTC, 217.182.129.124:6000,       2,   0.062,   0.055,   0.061,   0.054,   0.045,   0.062,   0.009
   2021-08-15 10:07:07.552151141 UTC, 217.182.129.124:6000,       3,   0.058,   0.057,   0.061,   0.055,   0.045,   0.062,   0.007
   2021-08-15 10:07:08.597797449 UTC, 217.182.129.124:6000,       4,   0.044,   0.054,   0.061,   0.053,   0.044,   0.062,   0.008
   2021-08-15 10:07:09.661213622 UTC, 217.182.129.124:6000,       5,   0.063,   0.057,   0.062,   0.055,   0.044,   0.063,   0.008
   2021-08-15 10:07:10.716040109 UTC, 217.182.129.124:6000,       6,   0.054,   0.056,   0.062,   0.055,   0.044,   0.063,   0.007
   2021-08-15 10:07:11.776319823 UTC, 217.182.129.124:6000,       7,   0.059,   0.057,   0.062,   0.055,   0.044,   0.063,   0.007
   2021-08-15 10:07:12.822832161 UTC, 217.182.129.124:6000,       8,   0.045,   0.056,   0.062,   0.055,   0.044,   0.063,   0.007
   2021-08-15 10:07:13.885131491 UTC, 217.182.129.124:6000,       9,   0.061,   0.057,   0.062,   0.055,   0.044,   0.063,   0.007


cardano-ping -c10 -m 8 -h relays-alonzo.poolunder.com -p 8900
194.163.158.97:8900 network rtt: 0.050
194.163.158.97:8900 handshake rtt: 0.063128494s
194.163.158.97:8900 Negotiated version NodeToNodeVersionV7 8 False
                           timestamp,                 host,  cookie,  sample,  median,     p90,    mean,     min,     max,     std
   2021-08-15 10:08:27.724228132 UTC,  194.163.158.97:8900,       0,   0.056,   0.056,   0.056,   0.056,   0.056,   0.056,     NaN
   2021-08-15 10:08:28.777609359 UTC,  194.163.158.97:8900,       1,   0.052,   0.054,   0.056,   0.054,   0.052,   0.056,   0.003
    2021-08-15 10:08:29.83503369 UTC,  194.163.158.97:8900,       2,   0.056,   0.055,   0.056,   0.055,   0.052,   0.056,   0.003
   2021-08-15 10:08:30.903684367 UTC,  194.163.158.97:8900,       3,   0.067,   0.056,   0.065,   0.058,   0.052,   0.067,   0.007
   2021-08-15 10:08:31.949762426 UTC,  194.163.158.97:8900,       4,   0.045,   0.055,   0.064,   0.055,   0.045,   0.067,   0.008
   2021-08-15 10:08:33.008356891 UTC,  194.163.158.97:8900,       5,   0.057,   0.056,   0.064,   0.056,   0.045,   0.067,   0.007
   2021-08-15 10:08:34.056491598 UTC,  194.163.158.97:8900,       6,   0.048,   0.055,   0.064,   0.054,   0.045,   0.067,   0.007
   2021-08-15 10:08:35.111596234 UTC,  194.163.158.97:8900,       7,   0.054,   0.055,   0.063,   0.054,   0.045,   0.067,   0.007
   2021-08-15 10:08:36.171830582 UTC,  194.163.158.97:8900,       8,   0.058,   0.056,   0.063,   0.055,   0.045,   0.067,   0.006
   2021-08-15 10:08:37.228618586 UTC,  194.163.158.97:8900,       9,   0.056,   0.056,   0.063,   0.055,   0.045,   0.067,   0.006

```

**Comment by coot** [comment]
@rdlrt is there anything else that's suspicious on `178.238.224.187:6003`?  I can confirm that a connection with it timeouts (using `cardano-ping`).  What's the memory or CPU usage, overall file descriptors used by the process vs the system limit, etc.? Is there a firewall on the way? It would be interesting to see OS information on incoming connection from the perspective of such a node.

**Comment by gitmachtl** [comment]
@coot cardano-ping is not closing the connection during a ping sequence like cncli is doing it. i used cncli for the test to simulate new node connections.

**Comment by rdlrt** [comment]
@coot - Memory usage has not crossed 20% for past 3.5 days, file descriptors is set to 20480, used by proc is 42 as per fdinfo and lsof. There is firewall but port 6003 is allowed , no other rules like rate limiting on this one - besides, the issue can be replicated from within the OS using `nc -v 127.0.0.1 6003` as well
0.1

**Comment by gitmachtl** [comment]
CLIO1 is still also having the same issues.

**Comment by coot** [comment]
@rdlrt could you run an experiment: run `tcpdump` on the host, and run `cardano-ping`, please share a strace log.

**Comment by rdlrt** [comment]
Tracing socket calls against `nc -v 127.0.0.1 6003` for node PID (bad node vs good node) using `strace`:
Bad node's :
```
[pid 3900562] 14:04:03.384093 recvfrom(31<socket:[138443245]>,  <unfinished ...>
[pid 3900562] 14:04:03.384746 recvfrom(31<socket:[138443245]>, "\202\0\31\1\321", 5, 0, NULL, NULL) = 5
[pid 3900562] 14:04:03.385858 recvfrom(31<socket:[138443245]>, 0x4205865ef0, 8, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
[pid 3900564] 14:04:03.387219 epoll_ctl(4<anon_inode:[eventpoll]>, EPOLL_CTL_MOD, 31<socket:[138443245]>, {EPOLLIN|EPOLLONESHOT, {u32=31, u64=100574768134094879}} <unfinished ...>
[pid 3900562] 14:04:03.387590 writev(31<socket:[138443245]>, [{iov_base=";k\16{\200\10\0\5", iov_len=8}, {iov_base="\202\1\31\1\321", iov_len=5}], 2) = 13
```
Good node:
```
[pid 3120588] 14:08:24.374056 recvfrom(49<socket:[129224416]>,  <unfinished ...>
[pid 3120588] 14:08:24.374621 recvfrom(49<socket:[129224416]>, "\0\31)\234", 4, 0, NULL, NULL) = 4
[pid 3120588] 14:08:24.374942 recvfrom(49<socket:[129224416]>,  <unfinished ...>
[pid 3120588] 14:08:24.375058 epoll_ctl(3<anon_inode:[eventpoll]>, EPOLL_CTL_MOD, 49<socket:[129224416]>, {EPOLLIN|EPOLLONESHOT, {u32=49, u64=314881850814234673}}) = 0
[pid 3120584] 14:08:24.375497 writev(49<socket:[129224416]>, [{iov_base="J\371j,\200\10\0\4", iov_len=8}, {iov_base="\1\31)\234", iov_len=4}], 2) = 12
```
Interesting to see `Resource temporarily unavailable message` on bad node

**Comment by coot** [comment]
Could you also grep the logs for `ErrorPolicyAcceptException` in the log ( `TraceErrorPolicy` is set to `true` in the [config](https://hydra.iohk.io/build/7189190/download/1/alonzo-purple-config.json)).

**Comment by rdlrt** [comment]
I didnt see matching exactly that string, but added grep for error (that matches any policy - EXCEPT 'Refused NodeToNodeV_7' since those are heaps and come from old nodes) [here](https://gist.githubusercontent.com/rdlrt/eeb6cf02306d593a340cdbfcd485cfb3/raw/341936b64e40a5673d6e12c00c9116fd1a4ea136/greperrornodeall.log)

**Comment by rdlrt** [comment]
Also, a tcptump (using filter on port `-A "dst port 6003"`) for new sample for a netcat connection attempt:
```
dropped privs to tcpdump
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
14:52:00.308926 IP vcs52.zcom.it.36589 > vmi441841.6003: Flags [P.], seq 3775833043:3775833056, ack 2142226393, win 501, options [nop,nop,TS val 2055712024 ecr 1278635054], length 13
E..Ag9@.8......i.......s............+v.....
z...L6l............*.
14:52:00.330653 IP vcs52.zcom.it.36589 > vmi441841.6003: Flags [.], ack 14, win 501, options [nop,nop,TS val 2055712046 ecr 1278645077], length 0
E..4g:@.8......i.......s...................
z...L6.U
14:52:00.764283 IP vps-ecb24a37.vps.ovh.net.33369 > vmi441841.6003: Flags [P.], seq 2899708677:2899708690, ack 4107910501, win 501, options [nop,nop,TS val 784270837 ecr 820725306], length 13
E..AP.@.3.`.6%.?.....Y.s.......e.....M.....
....0.F:.............
14:52:00.773034 IP vps-ecb24a37.vps.ovh.net.33369 > vmi441841.6003: Flags [.], ack 14, win 501, options [nop,nop,TS val 784270846 ecr 820735316], length 0
E..4P.@.3.`.6%.?.....Y.s.......r...........
....0.mT
14:52:01.605402 IP cm47-145.liwest.at.57311 > vmi441841.6003: Flags [S], seq 1905125765, win 64240, options [mss 1460,sackOK,TS val 1749274124 ecr 0,nop,wscale 7], length 0
E..<..@.8..f.!/........sq...........,..........
hC..........
14:52:01.637912 IP static.157.86.12.49.clients.your-server.de.41697 > vmi4418416003: Flags [P.], seq 2015087488:2015087501, ack 1330577948, win 501, options [nop,nop,TS val 2838550474 ecr 2964018172], length 13
E..A.m@.8...1.V........sx...OO......h'.....
.0....S............Sy
14:52:01.638583 IP 128.199.79.187.42977 > vmi441841.6003: Flags [P.], seq 3003407582:3003407595, ack 1277895115, win 1481, options [nop,nop,TS val 2128535248 ecr 2510901931], length 13
E..A.X@.8..2..O........s..\.L+!............
~.....R..............
14:52:01.647484 IP static.157.86.12.49.clients.your-server.de.41697 > vmi441841.6003: Flags [.], ack 14, win 501, options [nop,nop,TS val 2838550484 ecr 2964028183], length 0
E..4.n@.8...1.V........sx...OO.)....]6.....
.0....{.
14:52:01.794802 IP 128.199.79.187.42977 > vmi441841.6003: Flags [.], ack 14, win 1481, options [nop,nop,TS val 2128535405 ecr 2510912088], length 0
E..4.Y@.8..>..O........s..\.L+!.....a1.....
~..m..zX
14:52:02.876705 IP li1704-34.members.linode.com.34369 > vmi441841.6003: Flags [P.], seq 492452945:492452958, ack 190189238, win 501, options [nop,nop,TS val 2613586681 ecr 1398068360], length 13
E..A.M@.:..4.h^".....A.s.Z<Q.V.......[.....
..*.ST...8.@.........
14:52:03.113257 IP li1704-34.members.linode.com.34369 > vmi441841.6003: Flags [.], ack 14, win 501, options [nop,nop,TS val 2613586917 ecr 1398078598], length 0
E..4.N@.:..@.h^".....A.s.Z<^.V.............
..+.ST..
^C
11 packets captured
11 packets received by filter
0 packets dropped by kernel
```

**Comment by coot** [comment]
I think that's a system issue not a bug in the node, especially since no `SYNACK` arrives back when a `SYN` is sent.  Devops could not reproduce it, but if you'll find a way to reproduce it in reliable way we will investigate it further.

**Comment by rdlrt** [comment]
So we troubleshoot it in mainnet (with 1000 out of 10000 instead of 10 out of 100 sample)? :-\
These nodes did not (and still dont) have issue if using binaries before alonzo purple...

**Comment by coot** [comment]
I would still expect `SYNACK`, the `recvfrom(31<socket:[138443245]>, 0x4205865ef0, 8, 0, NULL, NULL)` errored, which seems like the first recv that we issue to receive mux header.

**Comment by gitmachtl** [comment]
> I think that's a system issue not a bug in the node, especially since no `SYNACK` arrives back when a `SYN` is sent. Devops could not reproduce it, but if you'll find a way to reproduce it in reliable way we will investigate it further.

You know that we have now at least 5 nodes in alonzo-purple that had this issue!

**Comment by coot** [comment]
I know, I haven't closed the issue, just removed `bug` label and asked a team mate to re-asses this on Monday.

**Comment by DanTup** [comment]
I don't know if it's useful info, but this issue has been seen on my node too (CODER). I'm using the official docker images (and have for Alonzo Blue, White and Purple, and I do for mainnet). They're all on the same Kubernetes host using the exact same configs (just with the image name changed and new config files downloaded - I'm not rebuilding anything, just deleting nodes, updating config, wiping the `db` folder and restarting) so the only difference is the container image/its contents.

**Comment by rdlrt** [comment]
> I would still expect `SYNACK`, the `recvfrom(31<socket:[138443245]>, 0x4205865ef0, 8, 0, NULL, NULL)` errored, which seems like the first recv that we issue to receive mux header.

Maybe you misread the log or messages, the SYN_SENT is sent by netcat/cardano-ping/cncli , and they time out while waiting for ACK - not the other way round. The strace log shared in gist for good & bad node is for node process, so what you say double's down imo about the issue being on node not responding to those requests 🙂

**Comment by kevinhammond** [comment]
To be clear - the team is treating this very seriously.  It's just not yet obvious what is causing this problem

**Comment by gitmachtl** [comment]
5 nodes not reachable again: HOMR1, CODER, ONXY, ASTOR and YOADA
![image](https://user-images.githubusercontent.com/47434720/129632651-dd599500-349c-4a1e-a174-5720ddbd6e98.png)

**Comment by kevinhammond** [comment]
We have investigated this issue and made some fixes in 1.29.0-rc1/rc2.  Please report any future occurrences to us.  As usual, we recommend that active stake pools are run behind a relay node.  This will significantly reduce the likelihood of observing this and other network-related issues.

Thank you again for raising this with us and helping us with the investigation.

**Comment by rdlrt** [comment]
The issue was replicated atleast once on rc1/rc2 and a special build is being run - Karl would be aware

**Comment by rdlrt** [comment]
Multiple nodes (myself, Max [POLY], Simon [QCPOL], Mendrinos [CROWD] ) in purple network using rc2 have hit this issue again.
In my case, I had 1.29 binary ready, nut didnt restart it - as the fix for this issue should be applicable on this version itself.

Hopefully marking latest release was an error (as it is followed by many for upgrading mainnet nodes)?

**Comment by rdlrt** [comment]
Looks like the instance on mine was on rc1 - as downloading from hydra build in latest PR of 1.29.0-rc2 did not include karl's fixes.

**Comment by hodlonaut** [comment]
@kevinhammond  @karknu 

This morning my Alonzo network node stopped responding to nc requests. It was built from rc2 branch of code on my Unbuntu server, after some early fixes by Karl:

cardano-node version
cardano-node 1.28.0 - linux-x86_64 - ghc-8.10
git rev cc78734d263d0eec2b12070380cdfea02a5a8342

The node was running for 9 days before this happened, has TraceLocalErrorPolicy and TraceErrorPolicy turned on but no TraceAcceptPolicy.

If I don't hear back in the next 6 hours from someone wanting to take a look at / poke this node I will restart it with 1.29.0 build and TraceAcceptPolicy turned on.

**Comment by karknu** [comment]
@hodlonaut Please share the logs with me.

**Comment by karknu** [comment]
The exact cause of this behavior has not been identified yet.
So far three bugs that could cause the node to stop accepting new connections has been found and fixed.
1.29.0 contains https://github.com/input-output-hk/ouroboros-network/pull/3314 and https://github.com/input-output-hk/ouroboros-network/pull/3305  .
1.30.0 will contain https://github.com/input-output-hk/ouroboros-network/pull/3335

**Comment by karknu** [comment]
I suspect that this issue has been fixed in https://github.com/input-output-hk/ouroboros-network/pull/3398

**Comment by gufmar** [comment]
Thank you, @karknu, for following this difficult to track and reproduce problem to resolution.

**Comment by dorin100** [comment]
do we still need to keep this issue open?

**Comment by karknu** [comment]
No.
Fixed in 1.30.0.
